pool.go

```diff
@@ -38,57 +41,66 @@ func (p *pool) loadPool() ([]string, error) {
 }
 
 type pool struct {
-	config    *Config
-	endpoints MemberList
-	logger    *logger
-	c         *hashring.HashRing
+	config *Config
+	logger *logger
+
+	started chan struct{} // started signals that we've already initialized the pool once with Saturn endpoints.
+	refresh chan struct{} // refresh is used to signal the need for doing a refresh of the Saturn endpoints pool.
+	done    chan struct{} // done is used to signal that we're shutting down the Saturn endpoints pool and don't need to refresh it anymore.
+
 	lk        sync.RWMutex
-	started   chan struct{}
-	refresh   chan struct{}
-	done      chan struct{}
+	endpoints MemberList         // guarded by lk
+	c         *hashring.HashRing // guarded by lk
 }
 
+// MemberList is the list of Saturn endpoints that are currently members of the Caboose consistent hashing ring
+// that determines which Saturn endpoint to use to retrieve a given CID.
 type MemberList []*Member
 
+// ToWeights returns a map of Saturn endpoints to their weight on Caboose's consistent hashing ring.
 func (m MemberList) ToWeights() map[string]int {
 	ml := make(map[string]int, len(m))
 	for _, mm := range m {
-		ml[mm.string] = mm.replication
+		ml[mm.url] = mm.replication
 	}
 	return ml
 }
 
+// Member is a Saturn endpoint that is currently a member of the Caboose consistent hashing ring.
 type Member struct {
-	string
-	sync.Mutex
+	lk sync.Mutex
+
+	url         string
 	lastUpdate  time.Time
 	replication int
 }
 
+var defaultReplication = 20
+
 func NewMember(addr string) *Member {
-	return &Member{addr, sync.Mutex{}, time.Now(), 20}
+	return &Member{url: addr, lk: sync.Mutex{}, lastUpdate: time.Now(), replication: defaultReplication}
 }
 
 func (m *Member) String() string {
-	return string(m.string)
+	return string(m.url)
 }
 
 func (m *Member) ReplicationFactor() int {
 	return m.replication
 }
 
-func (m *Member) Downvote(debounce time.Duration) (*Member, bool) {
+func (m *Member) Updatevote(debounce time.Duration, updateWeightF func(old int) int) (*Member, bool) {
 	// this is a best-effort. if there's a correlated failure we ignore the others, so do the try on best-effort.
-	if m.TryLock() {
+	if m.lk.TryLock() {
 		if time.Since(m.lastUpdate) > debounce {
 			// make the down-voted member
-			nm := NewMember(m.string)
-			nm.replication = m.replication / 2
-			m.lastUpdate = time.Now()
-			m.Unlock()
+			nm := NewMember(m.url)
+			nm.replication = updateWeightF(m.replication)
+			nm.lastUpdate = time.Now()
```

Comment:
We need to update the `lastUpdate` time on this newly constructed member and not on the old member as the pool will now be updated with this member and we want to retain the last updated time across weight updates.

Reply:
@willscott

Reply:
i think change it on both - if a couple things time out simultaneously against the previous member, we want only one of them to do the update re-weighting